Makefile: remove use of `cd` when running tests

eksperimental opened the pull request:
Elixir failing tests don't give proper path, because of `cd`,
from Makefile we are running
`cd lib/elixir && ../../bin/elixir -r "test/elixir/test_helper.exs" -pr "test/elixir/**/*_test.exs"`
and when a test fails, we get.
`test/mix/tasks/new_test.exs:28`
when the path it should be
`lib/mix/test/mix/tasks/new_test.exs:28`

```sh
  1) test Sample (VersionTest)
     test/elixir/version_test.exs:7
     Assertion with == failed
     code: 1 == 2
     lhs:  1
     rhs:  2
     stacktrace:
       test/elixir/version_test.exs:8
```

with this PR it reads:

```sh
  1) test Sample (VersionTest)
     lib/elixir/test/elixir/version_test.exs:7
     Assertion with == failed
     code: 1 == 2
     lhs:  1
     rhs:  2
     stacktrace:
       lib/elixir/test/elixir/version_test.exs:8
```

with this PR we get away without changing directory in order to get the proper path of
the tests being executed

eksperimental commented:
This only address the issues with the tests,
but a few other tasks change directory, so the situation would be similar.
I'm not sure if that will have side effects, but we can try,
from
`$(Q) cd lib/elixir && ../../$(ELIXIRC) "lib/kernel.ex" -o ebin;`
to
`$(Q) $(ELIXIRC) "lib/elixir/lib/kernel.ex" -o "lib/elixir/ebin";`

josevalim commented:
We need to leave `cd` because we also need to support running tests inside the project using mix. Unfortunately, we can only support one approach, because some tests need to assert the path too (and that's why you have failures in the build).

eksperimental commented:
@josevalim I have corrected the tests, and i ran a `make clean test` locally, without any failure. I have updated the commit, but I'm not sure how it will work, now that the PR is closed

josevalim commented:
It is not a matter of fixing the tests. The issue is that this change break some workflows. @ericmj, can you confirm your workflow would be broken with this PR? Because tests would always need to be run from root?

Btw, I was going to reopen this pull request but Github doesn't let me. :(

ericmj commented:
The only thing that changes is the paths in the output? If so, it's fine to just skip copying the `lib/APP/` part of the path. I don't really see the benefit of the changed path though because you cant do much with it without removing the added parts.

josevalim commented:
@ericmj are you running tests inside lib/ex_unit, lib/mix and so on? If you are, this patch will break it because tests will now always assume being run from root.

ericmj commented:
I didn't see the build broke so I didn't realize that changes to tests had to be made. In that case I am :-1:. It may also cause conflicting `tmp/` directories, I prefer the current separation.

eksperimental commented:
@josevalim I am seeing to load dynamically the path, to run the tests from a different location than root.
I'm not sure about the `tmp/` dir issue that @ericmj mentioned

eksperimental commented:
I created a new PR https://github.com/elixir-lang/elixir/issues/3535

Tests are passing from whenever you are calling them.

I have also improved the tests loading dynamically the line numbers for the function call and the module definition. that would have caused that editing the file of the test itself

@ericmj could you please elaborate on the `tmp/` dir issue. So i can look into it.

just tell me guys what you think of it

ericmj commented:
I don't really see the benefit except for display different paths. We are fixing the tests now but since the paths are not tested in CI outside the makefile and it is unusual to run the tests with mix so that will be tested less often.

tmp probably shouldn't be an issue since we use paths relative from source files.

eksperimental commented:
I think it would be convenient to have the exact file path with the line number, from the root directory,  you just pass it to your text editor from the command line, and you are editing the test failing.
And potentially you could pass it to `grep | xargs text_editor` and open all the files at once
